optimism-children-indexer/optimism-children-indexer-logic/src/services/layer3_indexer/indexer.rs

```diff
@@ -0,0 +1,278 @@
+                false
+            } else {
+                true
+            }
+        });
+    }
+
+    /// Spawn a new indexer task with delay
+    fn spawn_task(&mut self, config: Layer3Chain, delay: Duration) {
+        let chain_id = config.chain_id;
+
+        let handle = self.tasks.spawn(async move {
+            let chain_id = config.chain_id;
+            let task = Layer3IndexerTask::new(config);
+            let result = task.run_with_delay(delay).await;
+            (chain_id, result)
+        });
+
+        self.abort_handles.insert(chain_id, handle);
+    }
+
+    /// Attempt to respawn a task after completion or failure
+    /// Only respawns if the chain still exists and is enabled
+    fn try_respawn(&mut self, chain_id: i64, succeeded: bool) {
+        if let Some(chain) = self.chains.get(&chain_id).cloned() {
+            let delay = self.calculate_restart_delay(&chain, succeeded);
+
+            tracing::debug!(
+                chain_id = chain_id,
+                delay_secs = delay.as_secs(),
+                succeeded = succeeded,
+                "[{}] Scheduling task restart.",
+                chain.chain_name
+            );
+
+            self.spawn_task(chain, delay);
+        }
+    }
+
+    /// Calculate appropriate restart delay
+    fn calculate_restart_delay(&self, chain: &Layer3Chain, succeeded: bool) -> Duration {
+        if !succeeded {
+            return RESTART_DELAY_FAILING;
+        }
+
+        match (chain.l3_last_indexed_block, chain.l3_latest_block) {
+            (last, Some(latest)) if last >= latest => {
+                // Fully synced - check less frequently
+                RESTART_DELAY_SYNCED
+            }
+            (_, Some(_)) => {
+                // Behind - catch up
+                RESTART_DELAY_BEHIND
+            }
+            _ => {
+                // No block info - use default
+                RESTART_DELAY_BEHIND
+            }
+        }
+    }
+
+    /// Handle task completion
+    async fn handle_task_completion(
+        &mut self,
+        chain_id: i64,
+        result: Result<Layer3IndexerTaskOutput>,
+    ) {
+        // Always remove from active tasks first
+        self.abort_handles.remove(&chain_id);
+
+        let chain_name = self
+            .chains
+            .get(&chain_id)
+            .map(|config| config.chain_name.as_str())
+            .unwrap_or("unknown");
+
+        match result {
+            Ok(output) => {
+                // Handle task result
+                let (config, items) = output;
+
+                // Store indexed items
+                if let Err(err) = self.store_indexed_items(items).await {
+                    tracing::error!(err = ?err, "Failed to store indexed items in database.");
+                } else {
+                    // Update chain config
+                    self.update_chain_state(config.clone())
+                        .await
+                        .expect("Failed to update chain state.");
```

Comment:
Storing items & updating chain state must be done in a single database transaction. Otherwise if `store_indexer_items` succeeds, but `update_chain_state` fails or our app crashes before it's run, we may end up with duplicated items.